File: backend/app/schemas/capability.py
from pydantic import BaseModel, Field
from typing import List, Optional


class CapabilityEvidence(BaseModel):
    provider: str = ""
    repository: str = ""
    files: List[str] = Field(default_factory=list)
    depth: int = 0
    reason: str = ""


class IndustryExpectation(BaseModel):
    capability_id: str = ""
    title: str = ""
    detail: str = ""
    relevance: List[str] = Field(default_factory=list)


class CapabilityMatchedSignal(BaseModel):
    signal: str = ""
    statement: str = ""
    source_type: str = ""
    support: float = 0.0
    files: List[str] = Field(default_factory=list)


class CapabilityMissingSignal(BaseModel):
    signal: str = ""
    statement: str = ""


class CapabilityExplanation(BaseModel):
    summary: str = ""
    status_reason: str = ""
    evidence_strength: str = ""
    matched_signals: List[CapabilityMatchedSignal] = Field(default_factory=list)
    missing_signals: List[CapabilityMissingSignal] = Field(default_factory=list)


class CapabilityItem(BaseModel):
    id: str = ""
    title: str = ""
    summary: str = ""
    observable_abilities: List[str] = Field(default_factory=list)
    related_skills: List[str] = Field(default_factory=list)
    support: float = 0.0
    evidence: List[CapabilityEvidence] = Field(default_factory=list)
    # Interpretation layer (deterministic; support/evidence unchanged).
    status: str = ""
    knowledge_statement: str = ""
    # Phase 2: why this status was assigned (read-only explanation).
    capability_explanation: CapabilityExplanation = Field(
        default_factory=CapabilityExplanation
    )


class WhatInauraKnowsArea(BaseModel):
    capability_id: str = ""
    capability_title: str = ""
    statement: str = ""
    support: float = 0.0
    status: str = ""


class WhatInauraKnowsEvidenceSummary(BaseModel):
    evidence_count: int = 0
    source_types: List[str] = Field(default_factory=list)
    implementation_evidence_count: int = 0


class WhatInauraKnows(BaseModel):
    summary: str = ""
    demonstrated_areas: List[WhatInauraKnowsArea] = Field(default_factory=list)
    developing_areas: List[WhatInauraKnowsArea] = Field(default_factory=list)
    unverified_areas: List[WhatInauraKnowsArea] = Field(default_factory=list)
    evidence_summary: WhatInauraKnowsEvidenceSummary = Field(
        default_factory=WhatInauraKnowsEvidenceSummary
    )


class DemonstratedCapability(BaseModel):
    id: str = ""
    title: str = ""
    support: float = 0.0
    confidence: float = 0.0
    evidence: List[CapabilityEvidence] = Field(default_factory=list)


class MissingCapability(BaseModel):
    id: str = ""
    title: str = ""
    support: float = 0.0
    status: str = ""
    priority: float = 0.0
    priority_category: str = "covered"
    next_actions: List[str] = Field(default_factory=list)
    resources: List[dict] = Field(default_factory=list)
    evidence: List[CapabilityEvidence] = Field(default_factory=list)


class RequirementProvenance(BaseModel):
    source: str = ""
    source_url: Optional[str] = None
    source_quality: Optional[float] = None
    evidence_strength: str = ""
    required_level: float = 0.0
    importance: float = 0.0
    role_relevance: str = ""
    evidence_context: str = ""
    supporting_chunks: List[dict] = Field(default_factory=list)


class SkillCapability(BaseModel):
    skill: str
    slug: str = ""
    role: str = ""
    category: str = ""
    prerequisites: List[str] = Field(default_factory=list)
    proficiency: float = 0.0
    confidence: float = 0.0
    required_level: float = 0.0
    importance: float = 0.0
    status: str = ""
    priority: float = 0.0
    priority_category: str = "covered"
    industry_expectations: List[IndustryExpectation] = Field(default_factory=list)
    capabilities: List[CapabilityItem] = Field(default_factory=list)
    demonstrated_capabilities: List[DemonstratedCapability] = Field(default_factory=list)
    missing_capabilities: List[MissingCapability] = Field(default_factory=list)
    evidence_gap: List[str] = Field(default_factory=list)
    skill_gap: List[str] = Field(default_factory=list)
    evidence_sources: List[CapabilityEvidence] = Field(default_factory=list)
    requirement: RequirementProvenance = Field(default_factory=RequirementProvenance)
    explanation: str = ""
    what_inaura_knows: WhatInauraKnows = Field(default_factory=WhatInauraKnows)


class CapabilityMapResponse(BaseModel):
    role: str
    generated_at: str = ""
    capability_model_version: str = ""
    engine_version: str = ""
    skills: List[SkillCapability] = Field(default_factory=list)
    summary: dict = Field(default_factory=dict)

File: frontend/src/services/capability.ts
import { apiFetch } from "./api";

export type CapabilityEvidence = {
  provider: string;
  repository: string;
  files: string[];
  depth: number;
  reason: string;
};

export type IndustryExpectation = {
  capability_id: string;
  title: string;
  detail: string;
  relevance: string[];
};

export type CapabilityMatchedSignal = {
  signal: string;
  statement: string;
  source_type: string;
  support: number;
  files: string[];
};

export type CapabilityMissingSignal = {
  signal: string;
  statement: string;
};

export type CapabilityExplanation = {
  summary: string;
  status_reason: string;
  evidence_strength: string;
  matched_signals: CapabilityMatchedSignal[];
  missing_signals: CapabilityMissingSignal[];
};

export type CapabilityItem = {
  id: string;
  title: string;
  summary: string;
  observable_abilities: string[];
  related_skills: string[];
  support: number;
  evidence: CapabilityEvidence[];
  status?: string;
  knowledge_statement?: string;
  capability_explanation?: CapabilityExplanation;
};

export type WhatInauraKnowsArea = {
  capability_id: string;
  capability_title: string;
  statement: string;
  support: number;
  status: string;
};

export type WhatInauraKnows = {
  summary: string;
  demonstrated_areas: WhatInauraKnowsArea[];
  developing_areas: WhatInauraKnowsArea[];
  unverified_areas: WhatInauraKnowsArea[];
  evidence_summary: {
    evidence_count: number;
    source_types: string[];
    implementation_evidence_count: number;
  };
};

export type DemonstratedCapability = {
  id: string;
  title: string;
  support: number;
  confidence: number;
  evidence: CapabilityEvidence[];
};

export type CapabilityResource = {
  title: string;
  url: string;
  provider: string;
};

export type MissingCapability = {
  id: string;
  title: string;
  support: number;
  status: string;
  priority: number;
  priority_category: string;
  next_actions: string[];
  resources: CapabilityResource[];
  evidence: CapabilityEvidence[];
};

export type RequirementProvenance = {
  source: string;
  source_url: string | null;
  source_quality: number | null;
  evidence_strength: string;
  required_level: number;
  importance: number;
  role_relevance: string;
  evidence_context: string;
  supporting_chunks: Array<{
    chunk_id: string | null;
    role: string | null;
    topic: string | null;
    source: string | null;
    source_url: string | null;
  }>;
};

export type SkillCapability = {
  skill: string;
  slug: string;
  role: string;
  /** Requirement category, e.g. "Frontend" (optional: older backends omit it) */
  category?: string;
  /** Canonical prerequisite skill slugs, e.g. ["javascript", "html_css"] */
  prerequisites?: string[];
  proficiency: number;
  confidence: number;
  required_level: number;
  importance: number;
  status: string;
  priority: number;
  priority_category: string;
  industry_expectations: IndustryExpectation[];
  capabilities: CapabilityItem[];
  demonstrated_capabilities: DemonstratedCapability[];
  missing_capabilities: MissingCapability[];
  evidence_gap: string[];
  skill_gap: string[];
  evidence_sources: CapabilityEvidence[];
  requirement: RequirementProvenance;
  explanation: string;
  what_inaura_knows?: WhatInauraKnows;
};

export type CapabilityMap = {
  role: string;
  generated_at: string;
  capability_model_version: string;
  engine_version: string;
  skills: SkillCapability[];
  summary: {
    total_skills: number;
    by_status: Record<string, number>;
  };
};

export function getCapabilityMap(payload: { target_role?: string; skill?: string }) {
  return apiFetch<CapabilityMap>("/analysis/capability-map", {
    params: {
      ...(payload.target_role ? { target_role: payload.target_role } : {}),
      ...(payload.skill ? { skill: payload.skill } : {}),
    },
  });
}
